src/Oracle/Oracle.cs

```diff
@@ -0,0 +1,429 @@
+                var data = task.Tx.GetHashData();
+                task.Signs.Where(p => !Crypto.VerifySignature(data, p.Value, p.Key)).ForEach(p => task.Signs.Remove(p.Key, out _));
+                data = task.BackupTx.GetHashData();
+                task.BackupSigns.Where(p => !Crypto.VerifySignature(data, p.Value, p.Key)).ForEach(p => task.BackupSigns.Remove(p.Key, out _));
+                task.BackupSigns.TryAdd(oraclePub, backupSign);
+            }
+
+            if (task.Tx == null)
+            {
+                task.Signs.TryAdd(oraclePub, sign);
+                task.BackupSigns.TryAdd(oraclePub, backupSign);
+                return;
+            }
+
+            if (Crypto.VerifySignature(task.Tx.GetHashData(), sign, oraclePub))
+                task.Signs.TryAdd(oraclePub, sign);
+            else if (Crypto.VerifySignature(task.BackupTx.GetHashData(), backupSign, oraclePub))
+                task.BackupSigns.TryAdd(oraclePub, backupSign);
+            else
+                throw new RpcException(-100, "Invalid response transaction sign");
+
+            if (CheckTxSign(snapshot, task.Tx, task.Signs) || CheckTxSign(snapshot, task.BackupTx, task.BackupSigns))
+            {
+                FinishedCache.Add(requestId);
+                PendingQueue.Remove(requestId, out _);
+            }
+        }
+
+        private bool CheckTxSign(StoreView snapshot, Transaction tx, ConcurrentDictionary<ECPoint, byte[]> Signs)
+        {
+            ECPoint[] nodes = NativeContract.Designate.GetDesignatedByRole(snapshot, Role.Oracle);
+            int m = nodes.Length - (nodes.Length - 1) / 3;
+            if (Signs.Count >= m && tx != null)
+            {
+                var contract = Contract.CreateMultiSigContract(m, nodes);
+                ScriptBuilder sb = new ScriptBuilder();
+                foreach (var pair in Signs.OrderBy(p => p.Key))
+                {
+                    sb.EmitPush(pair.Value);
+                    if (--m == 0) break;
+                }
+                var idx = tx.GetScriptHashesForVerifying(snapshot)[0] == contract.ScriptHash ? 0 : 1;
+                tx.Witnesses[idx].InvocationScript = sb.ToArray();
+
+                Log($"Send response tx: responseTx={tx.Hash}");
+
+                System.Blockchain.Tell(tx);
+                return true;
+            }
+            return false;
+        }
+
+        public void OnTimer(object source, ElapsedEventArgs e)
+        {
+            List<ulong> outOfDate = new List<ulong>();
+            foreach (var task in PendingQueue)
+            {
+                var span = TimeProvider.Current.UtcNow - task.Value.Timestamp;
+                if (span > TimeSpan.FromSeconds(RefreshInterval) && span < TimeSpan.FromSeconds(RefreshInterval * 2))
+                {
+                    foreach (var account in Wallet.GetAccounts())
+                        if (task.Value.BackupSigns.TryGetValue(account.GetKey().PublicKey, out byte[] sign))
```

Comment:
We query request once and then send backup tx signatures on timeout. Do I understand correctly?

Reply:
Yes